core/trino-main/src/main/java/io/trino/execution/querystats/PlanOptimizersStatsCollector.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.querystats;
+
+import io.trino.spi.eventlistener.QueryPlanOptimizerStatistics;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.optimizations.PlanOptimizer;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+
+public class PlanOptimizersStatsCollector
+{
+    private final Map<Class<?>, QueryPlanOptimizerStats> stats = new HashMap<>();
+    private final int queryReportedRuleStatsLimit;
+
+    public PlanOptimizersStatsCollector(int queryReportedRuleStatsLimit)
+    {
+        this.queryReportedRuleStatsLimit = queryReportedRuleStatsLimit;
+    }
+
+    public void recordRule(Rule<?> rule, boolean invoked, boolean applied, long elapsedNanos)
+    {
+        if (invoked) {
+            stats.computeIfAbsent(rule.getClass(), (key) -> new QueryPlanOptimizerStats(key.getCanonicalName()))
+                    .record(elapsedNanos, applied);
+        }
+    }
+
+    public void recordOptimizer(PlanOptimizer planOptimizer, long duration)
+    {
+        stats.computeIfAbsent(planOptimizer.getClass(), (key) -> new QueryPlanOptimizerStats(key.getCanonicalName()))
+                .record(duration, true);
+    }
+
+    public void recordFailure(Object rule)
```

Comment:
Sorry, I missed this one earlier. Why is this `Object` and not `Rule<?>`. If the reason is because of `PlanOptimizer` vs `Rule` being unrelated, then let's introduce two overloads to make the code more type-safe.